x/wasm/keeper/msg_dispatcher.go

```diff
@@ -82,28 +82,28 @@ func (d MessageDispatcher) DispatchSubmessages(ctx sdk.Context, contractAddr sdk
 		}
 		// first, we build a sub-context which we can use inside the submessages
 		subCtx, commit := ctx.CacheContext()
-		em := sdk.NewEventManager()
-		subCtx = subCtx.WithEventManager(em)
+		subCtx = subCtx.WithEventManager(sdk.NewEventManager()) // discard any context events
 
 		// check how much gas left locally, optionally wrap the gas meter
 		gasRemaining := ctx.GasMeter().Limit() - ctx.GasMeter().GasConsumed()
 		limitGas := msg.GasLimit != nil && (*msg.GasLimit < gasRemaining)
 
-		var err error
-		var events []sdk.Event
-		var data [][]byte
+		var (
+			events []sdk.Event
+			data   [][]byte
+			err    error
+		)
 		if limitGas {
 			events, data, err = d.dispatchMsgWithGasLimit(subCtx, contractAddr, ibcPort, msg.Msg, *msg.GasLimit)
 		} else {
 			events, data, err = d.messenger.DispatchMsg(subCtx, contractAddr, ibcPort, msg.Msg)
 		}
 
+		events = filterEvents(events)
```

Comment:
I see how the usage gets simpler with only one path to handle.

But this puts a lot more work on the people writing handlers.
Since there is much less chance someone forks this dispatcher, than someone adds a new handler (for custom messages on any chain), I think this actually makes the code a bit more fragile, and some events may get silently dropped with hard-to-debug issues.